### PR TITLE
Reject HTTP/2 header values with invalid characters

### PR DESCRIPTION
**Pull request by chrisvest**
This PR builds upon #12755.

Motivation:
In https://datatracker.ietf.org/doc/html/rfc7540#section-10.3 it says that only certain characters are valid in a header value:

> Any request or response that contains a character not permitted
> in a header field value MUST be treated as malformed (Section 8.1.2.6).
> Valid characters are defined by the "field-content" ABNF rule in
> Section 3.2 of [RFC7230].

Modification:
Add a header value validation step to HpackDecoder.

Result:
Header values are now validated against the Section 10.3, etc. rules.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@ejona86 PTAL as well

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 1]
```diff
@@ -436,6 +484,43 @@ private static boolean contains(Http2Headers headers, CharSequence name) {
...
+            return true;
+        }
+        final byte[] array;
+        final int start;
+        if (value instanceof AsciiString) {
+            AsciiString ascii = (AsciiString) value;
+            array = ascii.array();
+            start = ascii.arrayOffset();
+        } else {
+            array = value.toString().getBytes(CharsetUtil.US_ASCII);
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:508]: hmm.. I wonder if it would be better to use `charAt(...)` in this case to reduce copy / allocation.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 2]
```diff
@@ -436,6 +484,43 @@ private static boolean contains(Http2Headers headers, CharSequence name) {
...
+        int length = value.length();
+        if (length == 0) {
+            return true;
+        }
+        final byte[] array;
+        final int start;
+        if (value instanceof AsciiString) {
+            AsciiString ascii = (AsciiString) value;
+            array = ascii.array();
+            start = ascii.arrayOffset();
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:506]: why not use `ByteProcessor` ?

File: codec-http2/src/test/resources/io/netty/handler/codec/http2/testdata/testStaticTableEntries.json [hunk 3]
```diff
@@ -62,7 +61,7 @@
         "8182 8384 8586 87 8f90",
         "9192 9394 9596 9798 999a 9b9c 9d9e 9fa0",
         "a1a2 a3a4 a5a6 a7a8 a9aa abac adae afb0",
-        "b1b2 b3b4 b5b6 b7b8 b9ba bbbc bd"
+        "b1b2 b3b4 b5b6 b7b8 babb bcbd"
```
Comment (normanmaurer) [inline codec-http2/src/test/resources/io/netty/handler/codec/http2/testdata/testStaticTableEntries.json:64]: why this change ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 2 above)
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:506]: We need to differentiate between first and following characters, so it'd have to be stateful, which means we have to allocate one every time.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 1 above)
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java]: I've added specialization for AsciiString and CharSequence, respectively, so we avoid allocating arrays for non-AsciiString inputs.

File: codec-http2/src/test/resources/io/netty/handler/codec/http2/testdata/testStaticTableEntries.json (hunk 3 above)
Comment (chrisvest) [reply codec-http2/src/test/resources/io/netty/handler/codec/http2/testdata/testStaticTableEntries.json:64]: We talked about this one here: https://github.com/netty/netty/pull/12755#discussion_r961426189

**Comment by chrisvest** [comment]
@normanmaurer Comments addressed.

**Review by normanmaurer — APPROVED** [review]

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 4]
```diff
@@ -480,6 +483,58 @@ private static boolean contains(Http2Headers headers, CharSequence name) {
...
+        final int start = value.arrayOffset();
+        byte b = array[start];
+        if (b < 0x21 || b == 0x7F) {
+            return false;
+        }
+        int length = value.length();
+        for (int i = start + 1; i < length; i++) {
+            b = array[i];
+            if (b < 0x20 && b != 0x09 || b == 0x7F) {
+                return false;
```
Comment (idelpivnitskiy) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:517]: IIUC, it will mark obs-text (`%x80-FF`) as invalid, while spec allow it. Consider: https://github.com/apple/servicetalk/blob/ca56585472e946e1bfe2f56ba7ceb2eac4f6653c/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyH2HeadersToHttpHeaders.java#L546-L549

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 5]
```diff
@@ -416,6 +416,9 @@ private static HeaderType validate(int streamId, CharSequence name,
             throw streamError(streamId, PROTOCOL_ERROR,
                     "Illegal value specified for the 'TE' header (only 'trailers' is allowed).");
         }
+        if (!isValidHeaderValue(value)) {
```
Comment (idelpivnitskiy) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:419]: I'm agreed that this validation is important and we should follow the spec. On the other hand, header values are tend to be quite long. Validating each value may be a significant performance hit. There are cases when communication is trusted between peers and such validation is not necessary. Can we consider an opt-out config options from this validation?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 4 above)
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:517]: That's basically the same logic. It says all controls, except for horizontal tab are invalid, and that DEL is invalid.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 6]
```diff
@@ -416,6 +416,9 @@ private static HeaderType validate(int streamId, CharSequence name,
...
+            throw streamError(streamId, PROTOCOL_ERROR, "Illegal header value given for header '%s'.", name);
```
Comment (idelpivnitskiy) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:420]: The header value can contain sensitive info, we can not include the whole value in the exception message. But without any hint, it's harder to debug what's wrong with the value. One of the most frequent case in my practice is to have `\n` at the end of the value. 
Consider including the hex value of the invalid byte (and maybe event its position) in the exception message. Lots of users will say "thank you" if they don't have to go through frame logging in order to infer what is wrong with the header value.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 5 above)
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:419]: We have a config for validating the headers. You want a separate one for header _value_ validation?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 4 above)
Comment (idelpivnitskiy) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:517]: This is true for `isValidHeaderValueCharSequence` because it uses `char b = value.charAt(0)`. 

For `AsciiString`, you operate with `byte`, which is negative for obs-text (`%x80-FF`). Please add a test to make sure `byte[] bytes = {(byte) 0x80, (byte) 0xFF}` is a valid header value for both paths: CharSequence and AsciiString.
Also, consider testing all scenarios for both CharSequence and AsciiString.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 5 above)
Comment (idelpivnitskiy) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:419]: Yes, validation of the header-name is more important and is used by many netty users by default (header names usually has more predictable length). If we add more validation for header values, we should let users preserve pre-existing behavior of validating only header names but not values.

Comment (idelpivnitskiy) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:419]: Actually, since the same validation rule applies for HTTP/2 and HTTP/1.x, have you considered moving this validation to the upper layer where it can be shared between two?

Comment (idelpivnitskiy) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:419]: ServiceTalk already does validation of the header values for both protocols, if netty also does it internally for HTTP/2 only without option to opt-out or share validation between HTTP/2 and HTTP/1.x, ServiceTalk users may end up doing expensive validation 2 times.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 4 above)
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:514]: Right you are. The signed bytes tricked me.

**Comment by normanmaurer** [comment]
@chrisvest let me know once this is ready for re-review

**Comment by chrisvest** [comment]
@normanmaurer This is ready for a second review now.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 7]
```diff
@@ -480,6 +487,66 @@ private static boolean contains(Http2Headers headers, CharSequence name) {
...
+        int length = value.length();
+        for (int i = start + 1; i < length; i++) {
+            b = array[i] & 0xFF;
+            if (b < 0x20 && b != 0x09 || b == 0x7F) {
+                throw illegalHeaderValue(streamId, name, b, i - start);
+            }
+        }
+    }
+
+    private static void verifyValidHeaderValueCharSequence(int streamId, CharSequence name, CharSequence value)
```
Comment (idelpivnitskiy) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:529]: This method isn't tested. For all test cases you created, consider running them for both `AsciiString` and a regular `String`.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 5 above)
Comment (idelpivnitskiy) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java]: I would leave it up to @normanmaurer. The benefit of doing this now instead of a separate PR is the API impact. For example, if you end up moving values validation to `DefaultHeaders` to share that between HTTP/2 and HTTP/1.X, you won't need any API changes for `DefaultHttp2HeadersDecoder` made in this PR.

Btw, moving this validation to `DefaultHeaders` has another benefit: users will be able to control it for inbound and outbound messages. In my practice, users frequently make mistakes in outbound header values, like adding `\n` at the end of the value (especially when the value is pulled from somewhere). When an illegal header sent, they see a weird error at transport level that doesn't describe the cause for request failure. Very hard to debug unless you have a validation before sending the request that describes where is the mistake.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java [hunk 8]
```diff
@@ -153,6 +184,15 @@ protected final boolean validateHeaders() {
         return validateHeaders;
     }
 
+    /**
+     * Determines if the header values should be validated as a result of the decode operation, in addition to the
+     * {@linkplain #validateHeaders() headers}.
+     * @return {@code true} if the header values should be validated as a result of the decode operation.
+     */
+    protected boolean validateHeaderValues() { // Not 'final' due to backwards compatibility.
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:192]: There could be sub-classes in the wild that define their own `validateHeaderValues`, so if we added a final method with the same signature, those classes would fail to compile.

[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:192]: Agreed, this is an interesting case. On the other hand, if someone added a method with the same signature but lower visibility, they will be broken anyway.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 9]
```diff
@@ -127,7 +127,13 @@ final class HpackDecoder {
      * This method assumes the entire header block is contained in {@code in}.
      */
     public void decode(int streamId, ByteBuf in, Http2Headers headers, boolean validateHeaders) throws Http2Exception {
-        Http2HeadersSink sink = new Http2HeadersSink(streamId, headers, maxHeaderListSize, validateHeaders);
+        decode(streamId, in, headers, validateHeaders, false);
+    }
+
+    public void decode(int streamId, ByteBuf in, Http2Headers headers, boolean validateHeaders,
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:133]: Then we should do that for all the public members

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 10]
```diff
@@ -480,6 +487,66 @@ private static boolean contains(Http2Headers headers, CharSequence name) {
...
+            verifyValidHeaderValueCharSequence(streamId, name, value);
+        }
+    }
+
+    private static void verifyValidHeaderValueAsciiString(int streamId, CharSequence name, AsciiString value)
+            throws Http2Exception {
+        final byte[] array = value.array();
+        final int start = value.arrayOffset();
+        int b = array[start] & 0xFF;
+        if (b < 0x21 || b == 0x7F) {
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:517]: Does it make any noticeable difference either way?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:192]: In these cases we follow what revapi does by default. Otherwise we won't have a way to expose this, which would be a broken symmetry.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 7 above)
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:516]: As we are in the decoder private APIs, and the decoder only produce AsciiString, we can alternatively pin that down in the API and just not generalize across all CharSequences.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/HpackDynamicTableTest.java [hunk 11]
```diff
@@ -23,12 +24,16 @@
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
 public class HpackDynamicTableTest {
+    AsciiString foo = AsciiString.cached("foo");
+    AsciiString bar = AsciiString.cached("bar");
+    AsciiString hello = AsciiString.cached("hello");
+    AsciiString world = AsciiString.cached("world");
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/HpackDynamicTableTest.java:30]:
```suggestion
    private final AsciiString foo = AsciiString.cached("foo");
    private final AsciiString bar = AsciiString.cached("bar");
    private final AsciiString hello = AsciiString.cached("hello");
    private final AsciiString world = AsciiString.cached("world");
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 12]
```diff
@@ -430,29 +431,26 @@ private static boolean isConnectionHeader(CharSequence name) {
...
-        if (len == 17) {
-            // Transfer-Encoding is more common, so check it first.
-            return contentEqualsIgnoreCase(name, HttpHeaderNames.TRANSFER_ENCODING);
+        switch (len) {
+            case 7: return contentEqualsIgnoreCase(name, HttpHeaderNames.UPGRADE);
+            case 10: return contentEqualsIgnoreCase(name, HttpHeaderNames.CONNECTION) ||
+                    contentEqualsIgnoreCase(name, HttpHeaderNames.KEEP_ALIVE);
+            case 16: return contentEqualsIgnoreCase(name, HttpHeaderNames.PROXY_CONNECTION);
+            case 17: return contentEqualsIgnoreCase(name, HttpHeaderNames.TRANSFER_ENCODING);
+            case 25: return contentEqualsIgnoreCase(name, HttpHeaderNames.UPGRADE_INSECURE_REQUESTS);
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:448]: Add a default branch as well... You can even just return false there.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:504]: honestly I think this will not make any diff.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 13]
```diff
@@ -634,15 +669,17 @@ public void appendToHeaderList(CharSequence name, CharSequence value) {
...
-                    previousType = validate(streamId, name, previousType, headers, value);
-                } catch (Http2Exception ex) {
-                    validationException = ex;
-                    return;
+            try {
+                if (validateHeaders) {
+                    previousType = validateHeader(streamId, name, previousType, headers, value);
+                }
+                if (validateHeaderValues) {
+                    validateValidHeaderValue(streamId, name, value);
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:677]: Not sure if it matters but should we set `previousType` to the old value again as `validateHeaders` might throw after we have update it.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoderTest.java [hunk 14]
```diff
@@ -175,6 +174,79 @@ public void decodingConnectionRelatedHeadersMustFailValidation() throws Exceptio
...
+    public void decodingInvalidHeaderValueMustFailValidation() throws Exception {
+        final DefaultHttp2HeadersDecoder decoder = new DefaultHttp2HeadersDecoder(true, true);
+
+        for (int illegalFirstChar = 0; illegalFirstChar < 0x21; illegalFirstChar++) {
+            verifyValidationFails(decoder, encode(b(":method"), b("GET"),
+                    b("test_header"), new byte[]{ (byte) illegalFirstChar, (byte) 'a' }));
+        }
+        verifyValidationFails(decoder, encode(b(":method"), b("GET"), b("test_header"), b("\u007Fa")));
+
+        for (int illegalSecondChar = 0; illegalSecondChar < 0x21; illegalSecondChar++) {
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoderTest.java:187]: nit: we could also generate these and pass as `param`. This way it is easier to re-run just the failed one.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoderTest.java [hunk 15]
```diff
@@ -175,6 +174,79 @@ public void decodingConnectionRelatedHeadersMustFailValidation() throws Exceptio
...
+            assertThat(headers.get("test_header")).isEqualToIgnoringCase("a\tb");
+        } finally {
+            ReferenceCountUtil.release(buf);
+        }
+    }
+
+    @Test
+    public void headerValuesAllowObsText() throws Exception {
+        final DefaultHttp2HeadersDecoder decoder = new DefaultHttp2HeadersDecoder(true);
+        for (int i = 0x80; i <= 0xFF; i++) {
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoderTest.java:226]: same comment about param

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 13 above)
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:677]: Moving value validation before the regular validation would solve that.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java [hunk 16]
```diff
@@ -70,16 +103,17 @@ public DefaultHttp2HeadersDecoder(boolean validateHeaders, long maxHeaderListSiz
      */
     public DefaultHttp2HeadersDecoder(boolean validateHeaders, long maxHeaderListSize,
                                       @Deprecated int initialHuffmanDecodeCapacity) {
-        this(validateHeaders, new HpackDecoder(maxHeaderListSize));
+        this(validateHeaders, false, new HpackDecoder(maxHeaderListSize));
     }
 
     /**
      * Exposed Used for testing only! Default values used in the initial settings frame are overridden intentionally
```
Comment (presenth7s) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:110]: lowercase

**Review by presenth7s — COMMENTED** [review]
* Exposed Used for testing only! Default values used in the initial settings frame are overridden intentionally

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java (hunk 16 above)
Comment (presenth7s) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:110]: uppercase to lowercase：Used

Comment (presenth7s) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:110]: uppercase to lowercase: Used

**Review by presenth7s — CHANGES_REQUESTED** [review]
Exposed used for testing only! Default values used in the initial settings frame are overridden intentionally

**Comment by chrisvest** [comment]
@normanmaurer @idelpivnitskiy I've consolidated the validation and pushed it down into the headers implementation, except for the validation that HTTP/2 headers are decoded pseudo-headers first, since I don't think we can compatibly place such a temporal restriction on the `Http2Headers` interface.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValidationUtil.java [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+        //  And: https://datatracker.ietf.org/doc/html/rfc5234#appendix-B.1
+        int length = value.length();
+        if (length == 0) {
+            return -1;
+        }
+        if (value instanceof AsciiString) {
+            return verifyValidHeaderValueAsciiString((AsciiString) value);
+        } else {
+            return verifyValidHeaderValueCharSequence(value);
+        }
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValidationUtil.java:124]:
```suggestion
        return verifyValidHeaderValueCharSequence(value);
```

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValidationUtil.java [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+        final byte[] array = value.array();
+        final int start = value.arrayOffset();
+        int b = array[start] & 0xFF;
+        if (b < 0x21 || b == 0x7F) {
+            return 0;
+        }
+        int length = value.length();
+        for (int i = start + 1; i < length; i++) {
+            b = array[i] & 0xFF;
+            if (b < 0x20 && b != 0x09 || b == 0x7F) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValidationUtil.java:137]: can you please add a link to the RFC for these values ?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValidationUtil.java [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+        } else {
+            return verifyValidHeaderValueCharSequence(value);
+        }
+    }
+
+    private static int verifyValidHeaderValueAsciiString(AsciiString value) {
+        final byte[] array = value.array();
+        final int start = value.arrayOffset();
+        int b = array[start] & 0xFF;
+        if (b < 0x21 || b == 0x7F) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValidationUtil.java:131]: can you please add a link to the RFC for these values ?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValidationUtil.java [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+            if (b < 0x20 && b != 0x09 || b == 0x7F) {
+                return i - start;
+            }
+        }
+        return -1;
+    }
+
+    private static int verifyValidHeaderValueCharSequence(CharSequence value) {
+        int b = value.charAt(0);
+        if (b < 0x21 || b == 0x7F) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValidationUtil.java:146]: can you please add a link to the RFC for these values ?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValidationUtil.java [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+            return 0;
+        }
+        int length = value.length();
+        for (int i = 1; i < length; i++) {
+            b = value.charAt(i);
+            if (b < 0x20 && b != 0x09 || b == 0x7F) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderValidationUtil.java:152]: can you please add a link to the RFC for these values ?

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java [hunk 22]
```diff
@@ -111,6 +144,7 @@ public DefaultHeaders(HashingStrategy<K> nameHashingStrategy,
         entries = new HeaderEntry[findNextPositivePowerOfTwo(max(2, min(arraySizeHint, 128)))];
         hashMask = (byte) (entries.length - 1);
         head = new HeaderEntry<K, V>();
+        this.valueValidator = valueValidator;
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java:147]: Missing null check

File: codec-http2/src/test/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandlerTest.java [hunk 23]
```diff
@@ -164,7 +164,6 @@ public void testHttpScheme() throws Exception {
         final Http2Headers http2Headers =
                 new DefaultHttp2Headers().method(new AsciiString("GET")).path(new AsciiString("/example"))
                         .authority(new AsciiString("www.example.org:5555")).scheme(new AsciiString("http"))
-                        .scheme(new AsciiString("http"))
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandlerTest.java:167]: why this change ?

File: codec-http2/src/test/java/io/netty/handler/codec/http2/HpackDynamicTableTest.java [hunk 24]
```diff
@@ -23,12 +24,16 @@
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
 public class HpackDynamicTableTest {
+    private final AsciiString foo = AsciiString.cached("foo");
+    private final AsciiString bar = AsciiString.cached("bar");
+    private final AsciiString hello = AsciiString.cached("hello");
+    private final AsciiString world = AsciiString.cached("world");
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/HpackDynamicTableTest.java:30]: nit: these could be static...

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java (hunk 22 above)
Comment (idelpivnitskiy) [reply codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java:147]: +1, consider verifying for null before any allocations

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java [hunk 25]
```diff
@@ -964,6 +1020,14 @@ protected ValueConverter<V> valueConverter() {
         return valueConverter;
     }
 
+    protected NameValidator<K> nameValidator() {
```
Comment (idelpivnitskiy) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java:1023]: I see that this is required for HTTP/2 headers impl. Instead of exposing these 2 methods as `protected` API consider caching the `nameValidator` in `DefaultHttp2Headers` instead of requiring this method. That way we can minimize public API surface.

**Review by idelpivnitskiy — COMMENTED** [review]
LGTM!

File: codec-http2/src/test/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandlerTest.java (hunk 23 above)
Comment (chrisvest) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandlerTest.java:167]: Mistook it for a broken test. Reverted.

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java (hunk 25 above)
Comment (chrisvest) [reply codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java:1023]: I think that these being hidden is inconsistent with the existing `valueConverter()`, and such inconsistencies lead to worse code. For instance, to not have this method available would mean duplicating the reference to the name validator, as you say. I think instead the `DefaultHeaders` class is designed to be extended, and these methods is part of making the design whole.

**Comment by chrisvest** [comment]
@normanmaurer @idelpivnitskiy Addressed your comments. Please take a look.

**Review by normanmaurer — APPROVED** [review]

**Comment by normanmaurer** [comment]
Ship it...

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java [hunk 26]
```diff
@@ -144,7 +144,7 @@ public DefaultHeaders(HashingStrategy<K> nameHashingStrategy, ValueConverter<V>
         entries = new HeaderEntry[findNextPositivePowerOfTwo(max(2, min(arraySizeHint, 128)))];
         hashMask = (byte) (entries.length - 1);
         head = new HeaderEntry<K, V>();
-        this.valueValidator = valueValidator;
+        this.valueValidator = checkNotNull(valueValidator, "valueValidator");
```
Comment (idelpivnitskiy) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java:147]: Consider moving the null check before allocating `entries` and `head`

**Review by idelpivnitskiy — APPROVED** [review]
🚢

File: codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java [hunk 27]
```diff
@@ -449,74 +384,16 @@ public CharSequence convertObject(Object value) {
...
-                        case '\t':
-                        case ' ':
-                            return 0;
-                        default:
-                            throw new IllegalArgumentException("only ' ' and '\\t' are allowed after '\\n'");
-                    }
-                default:
-                    break;
+        public void validate(CharSequence value) {
+            int index = HttpHeaderValidationUtil.validateValidHeaderValue(value);
```
Comment (valfirst) [inline codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java:392]: should the same validation rules be applied for both HTTP/1 and HTTP/2? This change introduces backward compatibility issues

Comment (idelpivnitskiy) [reply codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java:392]: @chrisvest this looks like omission. value validation should be off by default for all protocols with opt-in flag. I see that for `DefaultHttp2Headers` but not for `DefaultHttpHeaders
